core/store/src/trie/mem/mod.rs

```diff
@@ -1,8 +1,206 @@
+            heights: Default::default(),
+            shard_uid,
+        }
+    }
+
+    /// Inserts a new root into the trie. The given function should perform
+    /// the entire construction of the new trie, possibly based on some existing
+    /// trie nodes. This internally takes care of refcounting.
+    pub fn construct_root<Error>(
+        &mut self,
+        block_height: BlockHeight,
+        f: impl FnOnce(&mut Arena) -> Result<MemTrieNodeId, Error>,
+    ) -> Result<CryptoHash, Error> {
+        let root = f(&mut self.arena)?;
+        let state_root = root.as_ptr(self.arena.memory()).view().node_hash();
+        self.insert_root(state_root, root, block_height);
+        Ok(state_root)
+    }
+
+    fn insert_root(
+        &mut self,
+        state_root: StateRoot,
+        mem_root: MemTrieNodeId,
+        block_height: BlockHeight,
+    ) {
+        assert_ne!(state_root, CryptoHash::default());
+        let heights = self.heights.entry(block_height).or_default();
+        heights.push(state_root);
+        let new_ref = mem_root.add_ref(&mut self.arena);
+        if new_ref == 1 {
+            self.roots.entry(state_root).or_default().push(mem_root);
+        }
+        MEM_TRIE_NUM_ROOTS
+            .with_label_values(&[&self.shard_uid.shard_id.to_string()])
+            .set(self.roots.len() as i64);
+    }
+
+    /// Returns a root node corresponding to the given state root.
+    pub fn get_root<'a>(&'a self, state_root: &CryptoHash) -> Option<MemTrieNodePtr<'a>> {
+        assert_ne!(state_root, &CryptoHash::default());
+        self.roots.get(state_root).map(|ids| ids[0].as_ptr(self.arena.memory()))
+    }
+
+    /// Expires all trie roots corresponding to a height smaller than
+    /// `block_height`. This internally manages refcounts. If a trie root
+    /// is expired but is still used at a higher height, it will still be
+    /// valid until all references to that root expires.
+    pub fn delete_until_height(&mut self, block_height: BlockHeight) {
+        let mut to_delete = vec![];
+        self.heights.retain(|height, state_roots| {
+            if *height < block_height {
+                for state_root in state_roots {
+                    to_delete.push(*state_root)
+                }
+                false
+            } else {
+                true
+            }
+        });
+        for state_root in to_delete {
+            self.delete_root(&state_root);
+        }
+    }
+
+    fn delete_root(&mut self, state_root: &CryptoHash) {
+        if let Some(ids) = self.roots.get_mut(state_root) {
+            let last_id = ids.last().unwrap();
+            let new_ref = last_id.remove_ref(&mut self.arena);
+            if new_ref == 0 {
+                ids.pop();
+                if ids.is_empty() {
+                    self.roots.remove(state_root);
+                }
+            }
+        } else {
+            debug_assert!(false, "Deleting non-existent root: {}", state_root);
```

Comment:
My reasoning here is that the algorithm ensures that there's no way the user code can mess this up. The user code can only do construct_root(height, state root, fn) and delete_until_height(height). You can call these in any arbitrary manner and there's no way to introduce any inconsistency. So for any of these assertions to fail, it would mean we have a programming error within MemTries and that is critically bad, so I don't think the caller code should be burdened to handle an error result that is supposed to just be unreachable.

However, I could propagate more context so that if we do get a programming error we have a more useful panic message.